### PR TITLE
BUG: ETS multiplicative error models log-likelihood calculation

### PR DESCRIPTION
**Pull request by ltsaprounis**
Related to #7331 

**tl;dr:** This PR fixes a small issue in the likelihood calculation of ETS models with multiplicative errors, that caused them to have low ICs and unreasonable forecasts.

***

It appears that the problem raised in #7331 (ETS(MAM) giving unreasonable forecasts) is still there. The fix for the issue had the desired effect for the example time series in the issue but unfortunately the exploding forecasts can still appear for different time series like the one below:
<details><summary>Code to reproduce the plots</summary>

```python
from statsmodels.tsa.exponential_smoothing.ets import ETSModel
import numpy as np
from matplotlib import pyplot as plt

data = np.array([
    214307.1, 331110.3, 324617.4, 230818.8, 275551.5, 232633.5,
    281989.2, 245336.4, 296976. , 207572.7, 252719.4, 210795. ,
    142705.8, 146742.3, 261613.5, 276959.1, 365010. ,  48734.7,
    308919.9, 319401. , 150199.2, 103010.1, 109861.8,  89486.1,
    393645. , 421279.5, 347387.4, 269928. , 171278.7, 158568.9,
    280526.4, 195104.4, 230998.2, 226602.9, 267526.8, 258632.7
])

plt.plot(data)
plt.title("Original time series")
plt.show()

res = ETSModel(
    endog=data,
    seasonal_periods=12,
    error="mul",
    seasonal="mul",
    trend="add",
).fit()
pred = res.forecast(steps=12)
pred_fill = np.zeros_like(data)
pred_fill[:] = np.nan
plt.plot(data, label="endog")
plt.plot(np.concatenate([pred_fill, pred]), label="forecast")
plt.legend()
plt.title("ETS(MAM) forecast - statsmodels (main)")
plt.show()
```

</details>

image
image

Moreover, the log-likelihood of the unreasonable forecasts is unreasonably high making the AutoETS implementations (e.g. [sktime](https://github.com/sktime/sktime/blob/main/sktime/forecasting/ets.py), [aeon](https://github.com/aeon-toolkit/aeon/blob/main/aeon/forecasting/ets.py)) that are based on ETS model selection using information criteria (e.g. AIC) select the unreasonable forecasts.
&nbsp;
The problem comes from the Log-Likelihood calculation for ETS models with multiplicative errors, and more specifically the fail-safe mechanism for negative or zero yhat values here:
https://github.com/statsmodels/statsmodels/blob/23faea30e30ff759c3c9d3f1d57864b2aa68c827/statsmodels/tsa/exponential_smoothing/ets.py#L1162-L1169
&nbsp;
This way is not consistent with the literature (see screenshot below):
image
*screenshot from the [online version](https://www.openforecast.org/adam/ADAMETSEstimationLikelihood.html) of Svetunkov, I. (2023). Forecasting and Analytics with the Augmented Dynamic Adaptive Model (ADAM) (1st ed.). Chapman and Hall/CRC. https://doi.org/10.1201/9781003452652*
&nbsp;

In addition, it's not consistent with any other ETS implementations, where they all seem to follow the formula above for the last term of the log-likelihood, namely log(| yhat |). Below is the relevant code from the other OSS implementations I checked:
* [forecast](https://github.com/robjhyndman/forecast/blob/7d5a856e3ed6579bbd673234807e045c267f4521/src/etscalc.c#L106)
* [smooth](https://github.com/config-i1/smooth/blob/6ead0a2a3115075686b9cfe6240da63ee5db4418/R/adam.R#L2529)
* [nixtla - statsforecast](https://github.com/Nixtla/statsforecast/blob/9d75d22d21bec23063bc1df912cf7d415fd93fd6/statsforecast/ets.py#L96)

All the implementations listed above give similar Log-Likelihood values and AIC as well as reasonable forecasts. The AIC for ETS and this time series is best for the EST(ANN) model hence all the automatic model selection methods from the packages above select this one. Below are the forecasts from these different packages for ETS(MAM) for the same time "problematic" series:

<details><summary>(R) forecast </summary>

```R
library(forecast)
library(ggplot2)

data = ts(
  c(
    214307.1, 331110.3, 324617.4, 230818.8, 275551.5, 232633.5,
    281989.2, 245336.4, 296976. , 207572.7, 252719.4, 210795. ,
    142705.8, 146742.3, 261613.5, 276959.1, 365010. ,  48734.7,
    308919.9, 319401. , 150199.2, 103010.1, 109861.8,  89486.1,
    393645. , 421279.5, 347387.4, 269928. , 171278.7, 158568.9,
    280526.4, 195104.4, 230998.2, 226602.9, 267526.8, 258632.7
  ),
  freq=12
)

fcast_ets_mam = forecast::ets(data, model="MAM", ic="aic") %>% forecast(h=12)
autoplot(fcast_ets_mam)
summary(fcast_ets_mam)
```

</details>

image

<details><summary> (R) smooth </summary>

```R
library(smooth)
library(ggplot2)

data = ts(
  c(
    214307.1, 331110.3, 324617.4, 230818.8, 275551.5, 232633.5,
    281989.2, 245336.4, 296976. , 207572.7, 252719.4, 210795. ,
    142705.8, 146742.3, 261613.5, 276959.1, 365010. ,  48734.7,
    308919.9, 319401. , 150199.2, 103010.1, 109861.8,  89486.1,
    393645. , 421279.5, 347387.4, 269928. , 171278.7, 158568.9,
    280526.4, 195104.4, 230998.2, 226602.9, 267526.8, 258632.7
  ),
  freq=12
)

smooth_es_fit = smooth::es(data, model="MAM")
summary(smooth_es_fit)
smooth_es_fit %>% forecast(h=12) |> plot(main="Smooth ETS(MAM)")
```

</details>

image

<details><summary>(Python) nixtla - statsforecast</summary>

```python
from statsforecast.models import AutoETS as StatsforecastETS
import numpy as np
from matplotlib import pyplot as plt

data = np.array([
    214307.1, 331110.3, 324617.4, 230818.8, 275551.5, 232633.5,
    281989.2, 245336.4, 296976. , 207572.7, 252719.4, 210795. ,
    142705.8, 146742.3, 261613.5, 276959.1, 365010. ,  48734.7,
    308919.9, 319401. , 150199.2, 103010.1, 109861.8,  89486.1,
    393645. , 421279.5, 347387.4, 269928. , 171278.7, 158568.9,
    280526.4, 195104.4, 230998.2, 226602.9, 267526.8, 258632.7
])
s_fcaster = StatsforecastETS(model="MAM", season_length=12)
s_fcaster.fit(data)
pred = s_fcaster.predict(h=12)["mean"]
pred_fill = np.empty_like(data)
pred_fill[:] = np.nan
pred = np.concatenate([pred_fill, pred])
plt.plot(data, label="endog")
plt.plot(pred, label="forecast")
plt.title("Nixtla/statsforecast - ETS(MAM)")
plt.show()
print(s_fcaster.__dict__)
```

</details>

image

&nbsp;

This PR changes the code for the log-likelihood for multiplicative error models so that it's consistent with the literature and the other ETS implementations. This results in similar forecasts like above:

image

Here are the values for each the log-likelihood and the AIC for ETS(MAM) and that series for each implementation, including the change in statsmodels proposed in this PR.
| ETS Implementation  | Log-Likelihood | AIC |
| ------------- | ------------- | ------------- |
| **statsmodels - main**  | -79.417 | 194.835  |
| **(R) forecast**  | -470.094 |  974.188 |
| **(R) smooth**  | -453.326 |  936.652 |
| **nixtla - statsforecast**  | -467.193  |  968.386 |
| **statsmodels - this PR** | -453.251 |  942.503 |

***

**PS:** Many many many thanks to @config-i1 for helping me to figure this one out! 🙏

### CODE REVIEW

File: statsmodels/tsa/exponential_smoothing/ets.py [hunk 1]
```diff
@@ -1160,13 +1160,13 @@ def _loglike_internal(
...
-            # values so that the log-transformation yields very large negative
-            # values.
-            yhat[yhat <= 0] = 1 / (1e-8 * (1 + np.abs(yhat[yhat <= 0])))
-            logL -= np.sum(np.log(yhat))
+            # In some cases, yhat can become negative or zero, so that a
+            # multiplicative model is no longer well-defined. Zero values
+            # are replaced with 10^-32 (a very small number) and we take
+            # the absolute of yhat to avoid computing the log of negative
+            # numbers.
+            yhat[yhat == 0] = 1e-32
```
Comment (bashtage) [inline statsmodels/tsa/exponential_smoothing/ets.py:1168]: Really should leave the <=.  == 0 is likely too strict, and if it really is ==0, then <=0 will catch all cases.

File: statsmodels/tsa/exponential_smoothing/ets.py [hunk 2]
```diff
@@ -1160,13 +1160,13 @@ def _loglike_internal(
...
+            logL -= np.sum(np.log(np.abs(yhat)))
```
Comment (bashtage) [inline statsmodels/tsa/exponential_smoothing/ets.py:1169]: I don't think reflecting negatives is a good idea.  Better to replace with small numbers.

**Review by bashtage — CHANGES_REQUESTED** [review]
Good idea but can simplify a bit.

File: statsmodels/tsa/exponential_smoothing/ets.py (hunk 2 above)
Comment (config-i1) [reply statsmodels/tsa/exponential_smoothing/ets.py:1169]: Please note that the absolute value is not there to avoid negative numbers, but instead arises mathematically:

If $\epsilon_t \sim \mathcal{N}(0, \sigma^2)$ then $y_t = \mu_{y,t}(1+\epsilon_t) \sim \mathcal{N}(\mu_{y,t}, \mu_{y,t}^2 \sigma^2)$. We then end up with the following log-likelihood: 

$\ell(\boldsymbol{\theta}, {\sigma}^2 | \mathbf{y}) = -\frac{1}{2} \sum_{t=1}^T \log(2 \pi \mu_{y,t}^2 \sigma^2) -\frac{1}{2} \sum_{t=1}^T \frac{\epsilon_t^2}{\sigma^2}$ . As you see, we have $\mu_{y,t}^2$ in the formula above, which after simplifications can be written as: $-\frac{T}{2} \log(2 \pi \sigma^2) -\frac{1}{2} \sum_{t=1}^T \frac{\epsilon_t^2}{\sigma^2} - \sum_{t=1}^T \log |\mu_{y,t}|$.

So, it is necessary to have it there for the correct calculation of the log-likelihood. The derivations are based on the Section 11.1 here: https://openforecast.org/adam/ADAMETSEstimationLikelihood.html

Comment (ltsaprounis) [reply statsmodels/tsa/exponential_smoothing/ets.py:1169]: Thanks for reviewing the PR @bashtage and thanks for the additional explanation on the likelihood derivation @config-i1. 

Based on @config-i1's explanation and the other implementations, I think it makes sense to keep the `np.abs(yhat)` do you agree @bashtage?

**Comment by ltsaprounis** [comment]
> Please note that the absolute value is not there to avoid negative numbers, but instead arises mathematically:
> 
> If ϵt∼N(0,σ2) then yt=μy,t(1+ϵt)∼N(μy,t,μy,t2σ2). We then end up with the following log-likelihood:
> 
> ℓ(θ,σ2|y)=−12∑t=1Tlog⁡(2πμy,t2σ2)−12∑t=1Tϵt2σ2 . As you see, we have μy,t2 in the formula above, which after simplifications can be written as: −T2log⁡(2πσ2)−12∑t=1Tϵt2σ2−∑t=1Tlog⁡|μy,t|.
> 
> So, it is necessary to have it there for the correct calculation of the log-likelihood. The derivations are based on the Section 11.1 here: https://openforecast.org/adam/ADAMETSEstimationLikelihood.html

Made some changes to the comment for the log-likelihood calculation based on Ivan's comments above. Let me know what you think @bashtage.

File: statsmodels/tsa/exponential_smoothing/ets.py (hunk 1 above)
Comment (jseabold) [reply statsmodels/tsa/exponential_smoothing/ets.py:1169]: Fun corner cases. FWIW, I don't think you need the clipping at all anymore. It's almost certainly not being triggered here, and if you do switch it to clip the negatives, then it becomes unstable again.

**Comment by bashtage** [comment]
Thanks for this. It has been merged in. Never quite sure of the answer for this (y==0) but can give this a go and see how it works.
